File: it-agents/skills/code-architect/references/code-examples/java/Composite.java
// Composite — compose objects into trees, treat leaf and container uniformly.

import java.util.ArrayList;
import java.util.List;

interface Node {
    int size();
    String name();
}

class FileLeaf implements Node {
    private final String name;
    private final int bytes;
    FileLeaf(String name, int bytes) { this.name = name; this.bytes = bytes; }
    public int size() { return bytes; }
    public String name() { return name; }
}

class FolderComposite implements Node {
    private final String name;
    private final List<Node> children = new ArrayList<>();
    FolderComposite(String name) { this.name = name; }
    FolderComposite add(Node n) { children.add(n); return this; }
    public int size() {
        int total = 0;
        for (Node c : children) total += c.size();
        return total;
    }
    public String name() { return name; }
}

public class Composite {
    public static void main(String[] args) {
        FolderComposite root = new FolderComposite("root")
                .add(new FileLeaf("a.txt", 100))
                .add(new FolderComposite("sub")
                        .add(new FileLeaf("b.txt", 200))
                        .add(new FileLeaf("c.txt", 50)));
        System.out.println(root.name() + " total bytes: " + root.size());
    }
}

File: it-agents/skills/code-architect/references/code-examples/typescript/Composite.ts
// Composite — compose objects into trees, treat leaf and container uniformly.

interface FileSystemNode {
  name: string;
  size(): number;
}

class FileLeaf implements FileSystemNode {
  constructor(public name: string, private readonly bytes: number) {}

  size(): number {
    return this.bytes;
  }
}

class FolderNode implements FileSystemNode {
  private children: FileSystemNode[] = [];

  constructor(public name: string) {}

  add(node: FileSystemNode): this {
    this.children.push(node);
    return this;
  }

  // Aggregates children — same interface as a leaf.
  size(): number {
    return this.children.reduce((total, child) => total + child.size(), 0);
  }
}

function demo(): void {
  const root = new FolderNode("root")
    .add(new FileLeaf("readme.txt", 120))
    .add(
      new FolderNode("src")
        .add(new FileLeaf("index.ts", 400))
        .add(new FileLeaf("util.ts", 250)),
    );
  console.log(`total size of '${root.name}' = ${root.size()} bytes`);
}

demo();
